drivers/interrupt_controller/intc_nxp_pint.c

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2023, NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Based on STM32 EXTI driver, which is (c) 2016 Open-RnD Sp. z o.o. */
+
+#include <zephyr/device.h>
+#include <zephyr/irq.h>
+#include <errno.h>
+#include <zephyr/drivers/interrupt_controller/nxp_pint.h>
+
+#include <fsl_inputmux.h>
+
+#define DT_DRV_COMPAT nxp_pint
+
+#define PINT_NODE DT_INST(0, DT_DRV_COMPAT)
+
+static PINT_Type *pint_base = (PINT_Type *)DT_REG_ADDR(PINT_NODE);
+
+/* Describes configuration of PINT IRQ slot */
+struct pint_irq_slot {
+	nxp_pint_cb_t callback;
+	void *user_data;
+	uint8_t pin: 6;
+	uint8_t used: 1;
+};
+
+#define NO_PINT_ID 0xFF
+
+/* Tracks IRQ configuration for each pint interrupt source */
+static struct pint_irq_slot pint_irq_cfg[DT_PROP(PINT_NODE, num_lines)];
+/* Tracks pint interrupt source selected for each pin */
+static uint8_t pin_pint_id[DT_PROP(PINT_NODE, num_inputs)];
+
+#define PIN_TO_INPUT_MUX_CONNECTION(pin) \
+	((PINTSEL_PMUX_ID << PMUX_SHIFT) + (pin))
+
+/* Attaches pin to PINT IRQ slot using INPUTMUX */
+static void attach_pin_to_pint(uint8_t pin, uint8_t pint_slot)
+{
+	INPUTMUX_Init(INPUTMUX);
+
+	/* Three parameters here- INPUTMUX base, the ID of the PINT slot,
+	 * and a integer describing the GPIO pin.
+	 */
+	INPUTMUX_AttachSignal(INPUTMUX, pint_slot,
+			      PIN_TO_INPUT_MUX_CONNECTION(pin));
+
+	/* Disable INPUTMUX after making changes, this gates clock and
+	 * saves power.
+	 */
+	INPUTMUX_Deinit(INPUTMUX);
+}
+
+/**
+ * @brief Enable PINT interrupt source.
+ *
+ * @param pin: pin to use as interrupt source
+ *     0-64, corresponding to GPIO0 pin 1 - GPIO1 pin 31)
+ * @param trigger: one of nxp_pint_trigger flags
+ * @return 0 on success, or negative value on error
+ */
+int nxp_pint_pin_enable(uint8_t pin, enum nxp_pint_trigger trigger)
+{
+	uint8_t slot = 0U;
+
+	if (pin > ARRAY_SIZE(pin_pint_id)) {
+		/* Invalid pin ID */
+		return -EINVAL;
+	}
+	/* Find unused IRQ slot */
+	if (pin_pint_id[pin] != NO_PINT_ID) {
+		slot = pin_pint_id[pin];
+	} else {
+		for (slot = 0; slot < ARRAY_SIZE(pint_irq_cfg); slot++) {
+			if (!pint_irq_cfg[slot].used) {
+				break;
+			}
+		}
+		if (slot == ARRAY_SIZE(pint_irq_cfg)) {
+			/* No free IRQ slots */
+			return -EBUSY;
+		}
+		pin_pint_id[pin] = slot;
+	}
+	pint_irq_cfg[slot].used = true;
+	pint_irq_cfg[slot].pin = pin;
+	/* Attach pin to interrupt slot using INPUTMUX */
+	attach_pin_to_pint(pin, slot);
+	/* Now configure the interrupt. No need to install callback, this
+	 * driver handles the IRQ
+	 */
+	PINT_PinInterruptConfig(pint_base, slot, trigger, NULL);
+	return 0;
+}
+
+
+/**
+ * @brief disable PINT interrupt source.
+ *
+ * @param pin: pin interrupt source to disable
+ */
+void nxp_pint_pin_disable(uint8_t pin)
+{
+	uint8_t slot;
+
+	if (pin > ARRAY_SIZE(pin_pint_id)) {
+		return;
```

Comment:
nit but why not declare these functions to return int and return -EINVAL since that is essentially what you are checking for here

Reply:
Good question- I considered this, but I don't think we need to alert the user via a return value. In all the cases where the `disable` or `unset` functions returns prematurely, the action that the user requested is effectively not needed, since the pin is either invalid, or doesn't currently an interrupt enabled/callback installed. Therefore, the user doesn't really care *how* the pin's interrupt was disabled, since it still effectively will be.

If the user tries to use an invalid pin, they should find out about that in the `enable` function. I don't think the `disable` function needs to tell them that via a return code, since it shouldn't be called first anyways.